{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50f18ccd",
   "metadata": {},
   "source": [
    "# Looking at pig gut core species\n",
    "- run branchwater\n",
    "- what are core species\n",
    "- How much of the metaG do they make up when using new bins\n",
    "\n",
    "\n",
    "Ran sourmash gather using the new db titus made\n",
    "\n",
    "Link results to current folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51a8f9d7",
   "metadata": {},
   "source": [
    "## Check if there is pig or other contamination:\n",
    "- Create pangenome sketches at k51: Need a k51 db for this\n",
    "- Gather these against the euk db\n",
    "- Create sketches at k31 with just the CDS\n",
    "- Use these CDS sketches as a db for gather x the metagenomes\n",
    "- Most euk DNA non-coding.\n",
    "\n",
    "Exclude E. coli from this? Because getting the CDS from that is very large\n",
    "For now yes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c28f83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run checkM2 for single copy marker genes\n",
    "# redo with the correct genome set\n",
    "mamba activate checkm2\n",
    "\n",
    "srun --account=ctbrowngrp -p med2 -J checkm2 -t 6:00:00 -c 30 --mem=50gb --pty bash\n",
    "\n",
    "checkm2 predict --threads 30 --input MAGs --output-directory checkm2_result -x fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f9d12e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mamba create -n checkM python=3.9\n",
    "mamba install -c bioconda numpy matplotlib pysam\n",
    "mamba install -c bioconda hmmer prodigal pplacer\n",
    "pip3 install checkm-genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a281b46c",
   "metadata": {},
   "outputs": [],
   "source": [
    "export CHECKM_DATA_PATH=/home/amhorst/databases/checkm \n",
    "\n",
    "checkm taxon_set life Prokaryote Prokaryote_markerset\n",
    "\n",
    "checkm analyze --nt Prokaryote_markerset ../Floccifex_porci/MAGs/ Floccifex_porci -t 30 -x fasta \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0466dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try fetchMG (my preference tbh)\n",
    "/home/amhorst/.local/bin/fetchMGs extraction \\\n",
    "-t 30 /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/functional_profile_corespp/Floccifex_porci/MAGs/AtH2023_SRR5240748_MAG07.fasta \\\n",
    "\"genome\" ./test_MAG\n",
    "\n",
    "# run on all mags through a snakemake\n",
    "# snakefile\n",
    "srun --account=ctbrowngrp -p med2 -J fetchmg -t 2:00:00 -c 24 --mem=50gb --pty bash\n",
    "\n",
    "# snakemake\n",
    "snakemake --use-conda --resources mem_mb=50000 --rerun-triggers mtime \\\n",
    "-c 24 --rerun-incomplete -k -s snakefiles/function.smk -n\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96cf5348",
   "metadata": {},
   "outputs": [],
   "source": [
    "# symlink gather results\n",
    "ln -s /group/ctbrowngrp2/amhorst/2025-pigparadigm/results/gatherxgtdb+bins.species .\n",
    "\n",
    "# create a presence absence table from all gather results\n",
    "mamba activate pangenomics_dev \n",
    "\n",
    "sourmash scripts gather_tables \\\n",
    "gather/*.csv \\\n",
    "-c \"intersect_bp\" -p \\\n",
    "-o 251104_gatherout.species.pig.csv "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c38a1563",
   "metadata": {},
   "source": [
    "# run branchwater for all the core species in the db\n",
    "- Update the db to the most current one titus made, to get all the kmers for each species\n",
    "- Compare the new kmers to kmers gtdb\n",
    "- Compare % of pangenome kmers found in a metagenome\n",
    "- Branchwater cli \n",
    "\n",
    "## Also do a branchwter search with just the pangenome from GTDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a93e919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to make sketches of these first. \n",
    "# Both with all MAGs and only GTDB. So that we can compare. \n",
    "# snakefile\n",
    "srun --account=ctbrowngrp -p med2 -J branchwater -t 2:00:00 -c 12 --mem=50gb --pty bash\n",
    "\n",
    "# snakemake\n",
    "snakemake --use-conda --resources mem_mb=50000 --rerun-triggers mtime \\\n",
    "-c 12 --rerun-incomplete -k -s snakefiles/branchwater.smk -n\n",
    "\n",
    "# snakemake\n",
    "snakemake --use-conda --resources mem_mb=50000 --rerun-triggers mtime \\\n",
    "-c 12 --rerun-incomplete -k -s snakefiles/pangenome_sketch.smk -n\n",
    "\n",
    "\n",
    "snakemake --use-conda --resources mem_mb=20000 --rerun-triggers mtime \\\n",
    "-c 1 --rerun-incomplete -k -s snakefiles/branchwater.smk -n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d2ba54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do we have at k21? Needed for bw\n",
    "/home/ctbrown/scratch3/sourmash-midgie-raker/outputs.ath/rename/gtdb+bins.species.sig.zip\n",
    "\n",
    "gtdb+bins.species.k21.sig.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7798414",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do tax on these results ( use tax_metag.smk)\n",
    "srun --account=ctbrowngrp -p med2 -J tax -t 1:00:00 -c 40 --mem=20gb --pty bash\n",
    "\n",
    "mamba activate branchwater\n",
    "\n",
    "snakemake --use-conda --resources mem_mb=50000 --rerun-triggers mtime \\\n",
    "-c 40 --rerun-incomplete -k -s snakefiles/tax_metag.smk -n\n",
    "\n",
    "# results go into 251105_taxonomy.ipynb "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15d0f498",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print core microbes\n",
    "# see 251104_pick_core_spp.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e798c053",
   "metadata": {},
   "outputs": [],
   "source": [
    "# srun  for notebook \n",
    "srun --cpus-per-task=4 --mem=16G --time=04:00:00 --pty bash\n",
    "\n",
    "srun --account=ctbrowngrp -p med2 \\\n",
    "-J notebook -t 2:00:00 -c 2 --mem=16gb --pty bash\n",
    "\n",
    "mamba activate jupyter\n",
    "jupyter notebook --no-browser --port=8888 --ip=0.0.0.0\n",
    "\n",
    "# copy paste the link into the kernel selecter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "772330cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# spp function\n",
    "srun --account=ctbrowngrp -p med2 -J funprofiler -t 16:00:00 -c 4 --mem=70gb --pty bash\n",
    "\n",
    "# snakemake\n",
    "snakemake --use-conda --resources mem_mb=70000 --rerun-triggers mtime \\\n",
    "-c 4 --rerun-incomplete -k -s snakefiles/function.smk -n\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49419976",
   "metadata": {},
   "outputs": [],
   "source": [
    "# have a set of samples im pretty sure are not cow but pig gut. I want to gather x the euks\n",
    "# snakefile\n",
    "\n",
    "srun --account=ctbrowngrp -p med2 -J gather -t 6:00:00 -c 16 --mem=50gb --pty bash\n",
    "\n",
    "# snakemake\n",
    "snakemake --use-conda --resources mem_mb=50000 --rerun-triggers mtime \\\n",
    "-c 16 --rerun-incomplete -k -s snakefiles/test_gather.euk.smk -n\n",
    "\n",
    "# euk db \n",
    "/group/ctbrowngrp/sourmash-db/genbank-euks-2025.01/vertebrates.k51.sig.zip\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4166bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "srun --account=ctbrowngrp -p med2 -J funprofiler -t 4:00:00 -c 32 --mem=70gb --pty bash\n",
    "\n",
    "\n",
    "# move files that already have output\n",
    "for f in *.fasta; do\n",
    "    if [ -d \"../prokka/${f%.fasta}\" ]; then\n",
    "        mv \"$f\" done/\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8a5e3af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test code\n",
    "sourmash scripts abundhist \\\n",
    "<metagenome>.sig -I <pangenome.sig> \\\n",
    "--save-figure out.png\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/SRR5371371.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Prevotella_sp000434975/Prevotella_sp000434975.gtdb+mags.k31.zip \\\n",
    "--figure SRR5371371.x.Prevotella_sp000434975.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa789e20",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74f36ed0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1de2f3ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare to as deeply sequenced pig gut metaG SRR8960211 (49GB )\n",
    "# Cow one SRR5371371 = 45GB\n",
    "# and a shallower one ERR3211898  (4.6G)\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/SRR8960211.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Cryptobacteroides_sp000434935/Cryptobacteroides_sp000434935.gtdb+mags.k31.zip \\\n",
    "--figure SRR8960211.x.Cryptobacteroides_sp000434935.pig.png --bins 100\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/SRR5371371.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Cryptobacteroides_sp000434935/Cryptobacteroides_sp000434935.gtdb+mags.k31.zip \\\n",
    "--figure SRR5371371.x.Cryptobacteroides_sp000434935.png --bins 100\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/SRR8960211.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Bariatricus_sp004560705/Bariatricus_sp004560705.gtdb+mags.k31.zip \\\n",
    "--figure SRR8960211.x.Bariatricus_sp004560705.pig.png --bins 100\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/SRR5371371.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Bariatricus_sp004560705/Bariatricus_sp004560705.gtdb+mags.k31.zip \\\n",
    "--figure SRR5371371.x.Bariatricus_sp004560705.png --bins 100\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/SRR8960211.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Lactobacillus_amylovorus/Lactobacillus_amylovorus.gtdb+mags.k31.zip \\\n",
    "--figure SRR8960211.x.Lactobacillus_amylovorus.pig.png --bins 100\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/SRR5371371.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Lactobacillus_amylovorus/Lactobacillus_amylovorus.gtdb+mags.k31.zip \\\n",
    "--figure SRR5371371.x.Lactobacillus_amylovorus.png --bins 100\n",
    "\n",
    "sourmash scripts abundhist \\\n",
    "/group/ctbrowngrp/irber/data/wort-data/wort-sra/sigs/ERR3211898.sig \\\n",
    "-I /group/ctbrowngrp2/amhorst/2025-pig-corespp/results/pangenome/Lactobacillus_amylovorus/Lactobacillus_amylovorus.gtdb+mags.k31.zip \\\n",
    "--figure ERR3211898.x.Lactobacillus_amylovorus.png --bins 100"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (jupyter)",
   "language": "python",
   "name": "jupyter"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
