{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50f18ccd",
   "metadata": {},
   "source": [
    "# Looking at pig gut core species\n",
    "- run branchwater\n",
    "- what are core species\n",
    "- How much of the metaG do they make up when using new bins\n",
    "\n",
    "\n",
    "Ran sourmash gather using the new db titus made\n",
    "\n",
    "Link results to current folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96cf5348",
   "metadata": {},
   "outputs": [],
   "source": [
    "# symlink gather results\n",
    "ln -s /group/ctbrowngrp2/amhorst/2025-pigparadigm/results/gatherxgtdb+bins.species .\n",
    "\n",
    "# create a presence absence table from all gather results\n",
    "mamba activate pangenomics_dev \n",
    "\n",
    "sourmash scripts gather_tables \\\n",
    "gather/*.csv \\\n",
    "-c \"intersect_bp\" -p \\\n",
    "-o 251104_gatherout.species.pig.csv "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c38a1563",
   "metadata": {},
   "source": [
    "# run branchwater for all the core species in the db\n",
    "- Update the db to the most current one titus made, to get all the kmers for each species\n",
    "- Compare the new kmers to kmers gtdb\n",
    "- Compare % of pangenome kmers found in a metagenome\n",
    "- Branchwater cli \n",
    "\n",
    "## Also do a branchwter search with just the pangenome from GTDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a93e919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to make sketches of these first. \n",
    "# Both with all MAGs and only GTDB. So that we can compare. \n",
    "# snakefile\n",
    "srun --account=ctbrowngrp -p high2 -J branchwater -t 2:00:00 -c 12 --mem=50gb --pty bash\n",
    "\n",
    "# snakemake\n",
    "snakemake --use-conda --resources mem_mb=50000 --rerun-triggers mtime \\\n",
    "-c 12 --rerun-incomplete -k -s snakefiles/branchwater_and_hashes.smk -n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d2ba54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do we have at k21? Needed for bw\n",
    "/home/ctbrown/scratch3/sourmash-midgie-raker/outputs.ath/rename/gtdb+bins.species.sig.zip\n",
    "\n",
    "gtdb+bins.species.k21.sig.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7798414",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do tax on these results ( use tax_metag.smk)\n",
    "srun --account=ctbrowngrp -p med2 -J tax -t 1:00:00 -c 40 --mem=20gb --pty bash\n",
    "\n",
    "mamba activate branchwater\n",
    "\n",
    "snakemake --use-conda --resources mem_mb=50000 --rerun-triggers mtime \\\n",
    "-c 40 --rerun-incomplete -k -s snakefiles/tax_metag.smk -n\n",
    "\n",
    "# results go into 251105_taxonomy.ipynb "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15d0f498",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print core microbes\n",
    "# see 251104_pick_core_spp.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e798c053",
   "metadata": {},
   "outputs": [],
   "source": [
    "# srun  for notebook \n",
    "srun --cpus-per-task=4 --mem=16G --time=04:00:00 --pty bash\n",
    "\n",
    "srun --account=ctbrowngrp -p med2 \\\n",
    "-J notebook -t 2:00:00 -c 2 --mem=16gb --pty bash\n",
    "\n",
    "jupyter notebook --no-browser --port=8888 --ip=0.0.0.0\n",
    "\n",
    "# copy paste the link into the kernel selecter\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (jupyter)",
   "language": "python",
   "name": "jupyter"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
